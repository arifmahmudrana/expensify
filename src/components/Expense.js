import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const Expense = ({description, amount, createdAt, id}) => (
  <div>
    <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
    <p>{numeral(amount/100).format('$0,0.00')} - {moment(createdAt).format('Do MMM, YYYY')}</p>
  </div>
);

export default Expense;