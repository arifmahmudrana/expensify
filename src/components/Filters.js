import React, {Component} from 'react';
import {connect} from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setDate} from '../actions/filters';


export class Filters extends Component {
  state = {
    calendarFocused: null
  };

  onCalendarFocusChange = (calendarFocused) => this.setState(() => ({ calendarFocused }));

  render() {
    return (
      <div>
        <input type="text" value={this.props.text} onChange={e => this.props.setTextFilter(e.target.value)} />
        <select value={this.props.sortBy} onChange={e => this.props[`sortBy${e.target.value[0].toUpperCase()}${e.target.value.slice(1)}`]()}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.startDate}
          endDate={this.props.endDate}
          onDatesChange={this.props.setDate}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onCalendarFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
          startDateId="start"
          endDateId="end"
        />
      </div>
    );
  }
}

const mapStateToProps = ({filters}, props) => ({
  ...filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: value => dispatch(setTextFilter(value)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setDate: ({ startDate, endDate }) => dispatch(setDate(startDate, endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(Filters);