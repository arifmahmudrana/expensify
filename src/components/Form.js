import React, { Component } from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

class Form extends Component {
  constructor(props) {
    super(props)
    const {description, note, amount, createdAt, id} = this.getExpensesAttributes(props);
    this.state = {
      id,
      description,
      note,
      amount,
      createdAt: moment(createdAt),
      calendarFocused: false,
      error: ''
    };
  }
  getExpensesAttributes(props) {
    const {description = '', note = '', amount = '', createdAt, id = ''} = props;

    return {description, note, amount, createdAt, id};
  }
  componentDidUpdate(prevProps, prevState) {
    if (prevProps.id && this.props.id && prevProps.id !== this.props.id) {
      const {description, note, amount, createdAt, id} = this.getExpensesAttributes(this.props);
      this.setState(() => ({
        id,
        description,
        note,
        amount,
        createdAt: moment(createdAt)
      }));
    }
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }));
    }
  };
  onCreatedAtChange = createdAt => createdAt && this.setState(() => ({ createdAt }));
  onCalendarFocusChange = ({ focused }) => this.setState(() => ({ calendarFocused: focused }));
  onSubmit = (e) => {
    e.preventDefault();
    let error = '';
    if (!this.state.description || !this.state.amount) {
      error = 'Please provide description & amount'
    }

    this.setState(() => ({
      error
    }));

    if (!error) {
      const {description, amount, note, createdAt} = this.state
      this.props.onSubmit({
        description: description.trim(), 
        amount: parseFloat(amount, 10) * 100, 
        note: note.trim(), 
        createdAt: createdAt.valueOf()
      });
    }
  }
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onCreatedAtChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button type="submit">{this.props.id ? 'Edit' : 'Add'} Expense</button>
        </form>
      </div>
    );
  }
}

export default Form;