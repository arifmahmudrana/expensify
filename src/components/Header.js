import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink activeClassName="selected" to="/" exact>Home</NavLink>
    | <NavLink activeClassName="selected" to="/create">Create</NavLink> 
    | <NavLink activeClassName="selected" to="/help">Help</NavLink>
  </header>
);

export default Header;