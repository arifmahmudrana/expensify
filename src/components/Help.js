import React from 'react';

const Help = () => (
  <p>Help</p>
);

export default Help;