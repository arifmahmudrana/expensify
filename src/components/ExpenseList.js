import React from 'react';
import {connect} from 'react-redux';
import Expense from './Expense';
import selectorsExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length
      ? props.expenses.map(i => <Expense {...i} key={i.id} />)
      : <p>No expenses</p>
    }
  </div>
);

const mapStateToProps = ({expenses, filters}, props) => ({
  expenses: selectorsExpenses(expenses, filters)
});

export default connect(mapStateToProps)(ExpenseList);