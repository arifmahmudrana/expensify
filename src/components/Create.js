import React from 'react';
import Form from './Form';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

export const Create = ({addExpense}) => (
  <div>
    <h1>Add Expense</h1>
    <Form onSubmit={addExpense} />
  </div>
);

const mapDispatchToProps = (dispatch, {history}) => ({
  addExpense: expense => {
    dispatch(addExpense(expense));
    history.replace('/');
  }
});

export default connect(() => ({}), mapDispatchToProps)(Create);