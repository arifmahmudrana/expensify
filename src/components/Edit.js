import React from 'react';
import {connect} from 'react-redux';
import {editExpense, removeExpense} from '../actions/expenses';
import Form from './Form';

export const Edit = ({editExpense, removeExpense, expense}) => (
  <div>
    <h1>Edit Expense</h1>
    <Form onSubmit={editExpense} {...expense} amount={expense.amount/100} />
    <button onClick={removeExpense}>Delete</button>
  </div>
);

const mapDispatchToProps = (dispatch, {history, match}) => ({
  editExpense: expense => {
    dispatch(editExpense({...expense, id: match.params.id}));
    history.replace('/');
  },
  removeExpense() {
    dispatch(removeExpense(match.params.id));
    history.replace('/');
  }
});

const mapStateToProps = (state, ownProps) => ({
  expense: state.expenses.find(i => i.id === ownProps.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(Edit);