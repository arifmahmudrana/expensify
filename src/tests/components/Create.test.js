import React from 'react';
import { shallow } from 'enzyme';
import { Create } from '../../components/Create';
const expense = {
  description: 'Description 1',
  amount: 100,
  note: 'Note 1'
};

let wrapper, onSubmit;

beforeEach(() => {
  onSubmit = jest.fn();
  wrapper = shallow(<Create storeExpense={onSubmit} />);
});

test('create page should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should onSubmit be called', () => {
  wrapper.find('Form').prop('onSubmit')(expense);
  expect(onSubmit).toHaveBeenLastCalledWith(expense);
});