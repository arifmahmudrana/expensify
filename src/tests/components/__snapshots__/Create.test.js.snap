// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create page should render correctly 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <Form
    onSubmit={[MockFunction]}
  />
</div>
`;
